content-services/content-sources-backend
# Fixes 1468: add repo snapshot task

## Summary
This adds a task to snapshot a repository.  It is meant to recover from failures due to interruptions after:
* remote creation
* repo creation

but not due to interuptions during:
* repo sync
* publication creation
* distribution creation

This resiliency will have to be added during integration with the tasking system.

Other changes:
* Moves dao mocks out of test/mocks into the dao package.  This makes more sense from a dependency perspective
* introduces a Dao registry, to simplify creating and passing around these dao objects, especially in tests.
* Adds a snapshot dao with creation supported

## Testing steps
create a repository via ui or api:

```
POST http://{{host}}/api/content-sources/v1.0/repositories/
Content-Type: application/json
x-rh-identity: eyJpZGVudGl0eSI6eyJ0eXBlIjoiQXNzb2NpYXRlIiwiYWNjb3VudF9udW1iZXIiOiIxIiwiaW50ZXJuYWwiOnsib3JnX2lkIjoiMSJ9fX0K

{
  "name": "needed",
  "url": "https://jlsherrill.fedorapeople.org/fake-repos/needed-errata/",
  "distribution_versions": [
    "8"
  ],
  "distribution_arch": "x86_64"
}
```

run the 'task' manually:
```
go run cmd/external-repos/main.go pulp-create https://jlsherrill.fedorapeople.org/fake-repos/needed-errata/ 1
```

Confirm that a distribution was created:
```
curl http://localhost:8080/pulp/api/v3/distributions/rpm/rpm/  -u admin:password | jq
```

and you can fetch from it:

```
curl http://localhost:8080/pulp/content/a63e6d0a-9df9-4553-8eaa-e3c19180e15d/a4300e93-4b62-4704-aba5-13bfae95f593/repodata/repomd.xml
```






## TODO:
* [x] Down migration
